pkg/bindings/internal/util/tar_builder.go

```diff
@@ -0,0 +1,188 @@
+package util
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/containers/storage/pkg/fileutils"
+	"github.com/containers/storage/pkg/ioutils"
+	"github.com/hashicorp/go-multierror"
+	gzip "github.com/klauspost/pgzip"
+)
+
+type Devino struct {
+	Dev uint64
+	Ino uint64
+}
+
+type TarBuilder struct {
+	sources  []sourceMapping
+	excludes []string
+}
+
+type sourceMapping struct {
+	source string // Absolute path of the source directory/file
+	target string // Custom path inside the tar archive
+}
+
+// NewTarBuilder returns a new TarBuilder
+func NewTarBuilder() *TarBuilder {
+	return &TarBuilder{
+		sources:  []sourceMapping{},
+		excludes: []string{},
+	}
+}
+
+// Add adds a new source directory or file and the corresponding target inside the tar.
+func (tb *TarBuilder) Add(source string, target string) error {
+	absSource, err := filepath.Abs(source)
+	if err != nil {
+		return fmt.Errorf("failed to get absolute path for source: %v", err)
+	}
+	tb.sources = append(tb.sources, sourceMapping{source: absSource, target: target})
+	return nil
+}
+
+// Exclude adds patterns to be excluded during tar creation.
+func (tb *TarBuilder) Exclude(patterns ...string) {
+	tb.excludes = append(tb.excludes, patterns...)
+}
+
+// Build generates the tarball and returns a ReadCloser for the tar stream.
+func (tb *TarBuilder) Build() (io.ReadCloser, error) {
+	if len(tb.sources) == 0 {
+		return nil, fmt.Errorf("no source(s) added for tar creation")
+	}
+
+	pm, err := fileutils.NewPatternMatcher(tb.excludes)
+	if err != nil {
+		return nil, fmt.Errorf("processing excludes list %v: %w", tb.excludes, err)
+	}
+
+	pr, pw := io.Pipe()
+	gw := gzip.NewWriter(pw)
```

Comment:
This and tw feel like they could be in the `go func()` below - feels more natural to have them there, right next to the corresponding `defer Close()`

Reply:
👍